scripts/Diffraction/isis_powder/routines/sample_details.py

```diff
@@ -146,6 +195,42 @@ def thickness(self):
         else:
             raise RuntimeError("Thickness is not applicable for the shape type \"{}\"".format(self._shape_type))
 
+    def generate_sample_geometry(self):
+        """
+        Generates the expected input for sample geometry using the SampleDetails class
+        :param self: Instance of SampleDetails containing details about sample geometry and material
+        :return: A map of the sample geometry
+        """
+        return self._shape.generate_sample_geometry()
+
+    def generate_sample_material(self):
+        """
+        Generates the expected input for sample material using the SampleDetails class.
+        See SetSampleMaterial for documentation on this dictionary
+        :param self: Instance of SampleDetails containing details about sample geometry and material
+        :return: A map of the sample material
+        """
+        material_json = {'ChemicalFormula': self.material_object.chemical_formula}
+        if self.material_object.number_density:
+            material_json["SampleNumberDensity"] = self.material_object.number_density
+        if self.material_object.absorption_cross_section:
+            material_json["AttenuationXSection"] = self.material_object.absorption_cross_section
+        if self.material_object.scattering_cross_section:
+            material_json["ScatteringXSection"] = self.material_object.scattering_cross_section
+        return material_json
+
+    def generate_container_geometry(self):
+        if self._container_shape:
+            return self._container_shape.generate_container_geometry()
+        else:
+            return None
+
+    def generate_container_material(self):
+        if self.container_material_object:
+            return {'ChemicalFormula': self.container_material_object.chemical_formula}
+        else:
```

Comment:
If you run a reduction where the container material is a compound (not v realistic but say H2-O for example) you get an error from the SetSample algorithm:
```
The number density could not be determined. Please provide the number density, ZParameter and unit cell volume or mass density.
```
My sample_details object was set up like this:
```
sample_details = SampleDetails(width=4.0, height=4.0, thickness=0.1, center=[0, 0, 0], shape='slab', angle=0.0)
sample_details.set_material(chemical_formula='Si', number_density=2.0, crystal_density=1.0)
sample_details.set_container(front_thick=0.3175, back_thick=0.3175, chemical_formula='H2-O',number_density=1.0)
polaris.set_sample_details(sample=sample_details)
```


Don't think Mantid can look up the number density for compounds in the same way it can if the material is just an element.
I think you need to copy in something similar to lines 214-219 from generate_sample_material in here